File: Task3-4/Task4_py/Task4Monster.py
# /∗
# ∗ CSCI3180 Principles of Programming Languages 
# ∗
# ∗ --- Declaration --- 
# ∗
# ∗ I declare that the assignment here submitted is original except for source
# ∗ material explicitly acknowledged. I also acknowledge that I am aware of
# ∗ University policy and regulations on honesty in academic work, and of the
# ∗ disciplinary guidelines and procedures applicable to breaches of such policy
# ∗ and regulations, as contained in the website
# ∗ http://www.cuhk.edu.hk/policy/academichonesty/ 
# ∗
# ∗ Assignment 2
# ∗ Name : ZHANG Jingze
# ∗ Student ID : 1155107857
# ∗ Email Addr : jzzhang8@cse.cuhk.edu.hk 
# ∗/


from Monster import Monster

class Task4Monster(Monster):

	def __init__(self, MonsterID, healthCapacity):
		Monster.__init__(self, MonsterID, healthCapacity)

	def actionOnSoldier(self, soldier):
		if self.getHealth() <= 0:
			self.talk("You had defeated me.\n\n")
		else:
			if self.requireKey(soldier.getKeys()):
				self.fight(soldier)
			else:
				self.displayHints()

	def fight(self, soldier):
	    fightEnabled = True

	    while fightEnabled:
	      print("       | Monster{} | Soldier |".format(self.monsterID))
	      print("Health | {:>8d} | {:>7d} |\n".format(self.health, soldier.getHealth()))
	      print("=> What is the next step? (1 = Attack, 2 = Escape, 3 = Use Elixir.) Input: ",end="")

	      choice = input()

	      if choice == "1":
	        if self.loseHealth():
	          print("=> You defeated Monster{}.\n".format(self.monsterID))
	          self.dropItems(soldier)
	          soldier.addCoin()
	          fightEnabled = False
	        else:
	          if soldier.loseHealth():
	            self.recover(self.healthCapacity)
	            fightEnabled = False
	      elif choice == "2":
	        self.recover(self.healthCapacity)
	        fightEnabled = False
	      elif choice == "3":
	        if soldier.getNumElixirs() == 0:
	          print("=> You have run out of elixirs.\n")
	        else:
	          soldier.useElixir()
	      else:
	        print("=> Illegal choice!\n")

File: Task3-4/Task4_java/Task4Monster.java
import java.util.Scanner;


public class Task4Monster extends Monster {

	public Task4Monster(int monsterID, int healthCapacity){
		super(monsterID, healthCapacity);
	}

	public void actionOnSoldier(Task4Soldier soldier) {
		if (this.getHealth() <= 0) {
		  this.talk("You had defeated me.%n%n");
		} else {
		  if (this.requireKey(soldier.getKeys())) {
		    this.fight(soldier);
		  } else {
		    this.displayHints();
		  }
		}
	}


  public void fight(Task4Soldier soldier) {
    boolean fightEnabled = true;

    while (fightEnabled) {
      System.out.printf("       | Monster%d | Soldier |%n", this.getMonsterID());
      System.out.printf("Health | %8d | %7d |%n%n", this.getHealth(), soldier.getHealth());
      System.out.printf("=> What is the next step? (1 = Attack, 2 = Escape, 3 = Use Elixir.) Input: ");

      Scanner sc = new Scanner(System.in);

      String choice = sc.nextLine();

      if (choice.equalsIgnoreCase("1")) {
        if (this.loseHealth()) {
          System.out.printf("=> You defeated Monster%d.%n%n", this.getMonsterID());
          this.dropItems(soldier);
          soldier.addCoin();
          fightEnabled = false;
        } else {
          if (soldier.loseHealth()) {
            this.recover(this.getHealthCapacity());
            fightEnabled = false;
          }
        }
      } else if (choice.equalsIgnoreCase("2")) {
        this.recover(this.getHealthCapacity());
        fightEnabled = false;
      } else if (choice.equalsIgnoreCase("3")) {
        if (soldier.getNumElixirs() == 0) {
          System.out.printf("=> You have run out of elixirs.%n%n");
        } else {
          soldier.useElixir();
        }
      } else {
        System.out.printf("=> Illegal choice!%n%n");
      }
    }
  }


}





